### PR TITLE
e2e, sriov, Expose flawed mapping with test

### PR DESCRIPTION
**Pull request by RamLavi**
**What this PR does / why we need it**:
current sriov mapping is flawed on vmi's with more than 2 SRIOV
networks [0]. Adding a Xfail test to record it.

[0] https://github.com/kubevirt/kubevirt/issues/7444

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: tests/network/sriov.go
Comment (oshoval) [inline tests/network/sriov.go:676]: aint it `getInterfaceNameByMAC` ?

Comment (RamLavi) [reply tests/network/sriov.go:676]: The naming of these things are pretty lousy IMO, but 
`getInterfaceNameByMAC` returns the network name as it is on the vmi.spec.networks[0]Name
`getGuestInterfaceNameByMAC` returns the interface name as it looks inside the vmi guest (eth0, eth1, ...)
Although the both could be considered as returning the `InterfaceName` they return completely different things, so I don't see a reason to combine the 2 functions into 1.

Comment (oshoval) [reply tests/network/sriov.go:676]:
```
func getInterfaceNameByMAC(vmi *v1.VirtualMachineInstance, mac string) (string, error) {
	for _, iface := range vmi.Status.Interfaces {
		if iface.MAC == mac {
			return iface.InterfaceName, nil
		}
	}
	return "", fmt.Errorf("could not get sriov interface by MAC: no interface on VMI %s with MAC %s", vmi.Name, mac)
}
```

It returns `InterfaceName` as well (it is not `getInterfaceNetworkNameByMAC`)
isn't it ?

File: tests/network/sriov.go
Comment (ormergi) [inline tests/network/sriov.go:555]: Why are these checks necessary? 
this test is about checking that each SR-IOV interface is configured with the correct PCI address, isn't it?
If so I would skip them.

File: tests/network/sriov.go
Comment (ormergi) [inline tests/network/sriov.go:534]: How about: "should create a virtual machine with **SR-IOV** interfaces with custom MAC addresses referring **to the correct PCI address**"?

File: tests/network/sriov.go
Comment (ormergi) [inline tests/network/sriov.go:652]: nit: I would change the token to `PCI_SLOT_NAME=%s` to improve visibility, not a must though.

File: tests/network/sriov.go
Comment (RamLavi) [reply tests/network/sriov.go:555]: Why not check the usual things that are checked in this lane?

File: tests/network/sriov.go
Comment (EdDev) [inline tests/network/sriov.go:558]: "not working correctly"

I think you should wrap it over the the `pciAddressExists...` assertion, being specific to what may get wrong.

File: tests/network/sriov.go
Comment (EdDev) [inline tests/network/sriov.go:559]: I agree with @oshoval that you should use `getInterfaceNameByMAC` instead.
It will also be correct to check it was found, because if it was not, the next command will blow up for the wrong reason.

**Review by EdDev — CHANGES_REQUESTED** [review]
> e2e, sriov, Add xfail to track failing mappping

I think that the main focus of this commit is about exposing the flawed mapping.
The xfail is only the means to write the test, expressing it is not yet fixed in the production code.

I think you should mention xfail only in the body of the commit message, as a side note, nothing more (I think you actually did it already).

File: tests/network/sriov.go
Comment (RamLavi) [reply tests/network/sriov.go:652]: you mean this?
```suggestion
	command := fmt.Sprintf("grep -q %s /sys/class/net/%s/device/uevent\n", pciSlotName, interfaceName)
```

Comment (EdDev) [reply tests/network/sriov.go:652]: I think he means that given data as follows:
```
$ cat /sys/class/net/wlp0s20f3/device/uevent 
DRIVER=iwlwifi
PCI_CLASS=28000
PCI_ID=8086:06F0
PCI_SUBSYS_ID=8086:0070
PCI_SLOT_NAME=0000:00:14.3
MODALIAS=pci:v00008086d000006F0sv00008086sd00000070bc02sc80i00
```
To grep `PCI_SLOT_NAME=0000:00:14.3`.

File: tests/network/sriov.go
Comment (RamLavi) [reply tests/network/sriov.go:676]: ah, I thought you were talking about a different function `getInterfaceNetworkNameByMAC` .. yes you are totally right!! thanks :)

File: tests/network/sriov.go
Comment (RamLavi) [reply tests/network/sriov.go:555]: ACK I'll make test more specific to our case.

File: tests/network/sriov.go
Comment (RamLavi) [reply tests/network/sriov.go:534]: Done

File: tests/network/sriov.go
Comment (RamLavi) [reply tests/network/sriov.go:539]: true, we just need to mind the default interface. Done

File: tests/network/sriov.go
Comment (RamLavi) [reply tests/network/sriov.go:552]: moved to a more specific testing. removed

File: tests/network/sriov.go
Comment (RamLavi) [reply tests/network/sriov.go:558]: agree, changed

File: tests/network/sriov.go
Comment (RamLavi) [reply tests/network/sriov.go:559]: Ack, will also add an Eventually as the interfaceName does not appear at first

Comment (oshoval) [reply tests/network/sriov.go]: this one Poll `findIfaceByMAC`

**Comment by RamLavi** [comment]
fixed [here](https://github.com/kubevirt/kubevirt/compare/558d06cfeb625474e5dc2a882302d4b5191d5fb9..277c24ee063858172d470b3fee51a217d5938510)

File: tests/network/sriov.go
Comment (ormergi) [inline tests/network/sriov.go:547]: I think we need to check if the interface is of SR-IOV type, looks like it failed due to not finding the MAC address probably for the primary NIC 
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/8218/pull-kubevirt-e2e-kind-1.22-sriov/1554042038645690368

File: tests/network/sriov.go
Comment (ormergi) [inline tests/network/sriov.go:553]: The correct issue is https://github.com/kubevirt/kubevirt/issues/7444

File: tests/network/sriov.go
Comment (ormergi) [inline tests/network/sriov.go:551]: I think that the way errors are handled inside the Eventually func is confusing:
line 549 the error is asserted (and blow up the test).
line 551 the error is not asserted (returned instead).
Is it possible to handle both (and other) errors the same way?

Comment (oshoval) [reply tests/network/sriov.go:551]: I suggest to take a look please on
```
					ifaceName, err := findIfaceByMAC(virtClient, vmi, mac, 30*time.Second)
					Expect(err).NotTo(HaveOccurred())
```
It has the Eventually in it already and then this block won't be needed
(at least some of it)

**Comment by rmohr** [comment]
/hold

xfail was not helpful at all so far. Please disable tests or provide a fix instead. Almost 100% of xfail marked tests just failed for months, making it hard to see real issues.

File: tests/network/sriov.go
Comment (RamLavi) [reply tests/network/sriov.go:547]: Added.

**Comment by EdDev** [comment]
> /hold
> 
> xfail was not helpful at all so far. Please disable tests or provide a fix instead. Almost 100% of xfail marked tests just failed for months, making it hard to see issues.

Well, I think this is a opinionated matter and as a user of xfail, I disagree.
xfail is a well established practice to declare a tests as "expected to fail" and it has served well so far. It is currently serving the sig-network and not the whole project, so I do not see how it can bother the project as a whole.

As this has been already accepted by the network maintainer long ago, and there was no agreement to remove it, I see no particular reason to block this PR over this.
If you do have strong feeling over this, I think it is more appropriate to negotiate it with the SIG maintainer over blocking a contribution.

If a decision will be taken to remove xfail, then this one can be taken out as well.

/unhold

File: tests/network/sriov.go
Comment (RamLavi) [reply tests/network/sriov.go]: couldn't agree more. thanks!

**Comment by rmohr** [comment]
> Well, I think this is a opinionated matter and as a user of xfail, I disagree. xfail is a well established practice to declare a tests as "expected to fail" and it has served well so far. It is currently serving the sig-network and not the whole project, so I do not see how it can bother the project as a whole.

Every user who is looking on CI is  affected by it. There are tons of output on each run. You got a chance for more than one year to prove that it is worth having it. It clearly failed to provide value. You were the only one who wanted it and pushed for having it. Please remove it.

/hold

> 
> As this has been already accepted by the network maintainer long ago, and there was no agreement to remove it, I see no particular reason to block this PR over this. If you do have strong feeling over this, I think it is more appropriate to negotiate it with the SIG maintainer over blocking a contribution.
> 
> If a decision will be taken to remove xfail, then this one can be taken out as well.

**Comment by rmohr** [comment]
@EdDev you know that you are repeatedly pushing for changes against the maintainers wish. If we  then allow you to haven them, only to have extra difficulty removing them afterwards because it is then a project decision (which it was not at the first place), I fear we have to reduce your freedom on the project a little bit.

**Comment by EdDev** [comment]
> Every user who is looking on CI is affected by it.

AFAIK the incident in which one will look at the logs is when a network job fails.
Hopefully this is not happening a lot, we try to maintain the network tests without flakes.
I also think that my colleges and I (in the sig-network) look at the logs more than others, so it seems only fair IMO to consider our opinion on this matter.

> There are tons of output on each run.

Right, I agree. I just think the output is polluted by other things that are not useful, not by this xfail thing.
I am all in favor of improving the output log.
I would like to find the solution that does that and still preserve what I believe to be a useful functionality.

> You got a chance for more than one year to prove that it is worth having it. It clearly failed to provide value. You were the only one who wanted it and pushed for having it. Please remove it.

I actually think it was proven to provide value with the latest k8s releases.
Several tests, marked as xfail, stopped xfail-ing, which showed the functionality is now supported on the tested k8s version.
We should periodically examine why some tests are still not passing, to re-assure the scenario tested is something we are expecting to work.

**Comment by EdDev** [comment]
> @EdDev you know that you are repeatedly pushing for changes against the maintainers wish.

My contributions to the project are subject to the same rules: Getting LGTM from reviewers and approve from at least one maintainer that is willing to support the change.
I do not think that generalizing my contributions as being "against the maintainers wish" is correct, especially if most do get in.
 
> If we then allow you to haven them, only to have extra difficulty removing them afterwards because it is then a project decision (which it was not at the first place), I fear we have to reduce your freedom on the project a little bit.
 Feel free to bring this up to other maintainers, but note that going up the Red Hat manager chain will probably not help you.

I would prefer the discussion to focus on the technical details and not drift to personal matters.
I do not understand how my employer has anything to do with the subject this PR covers or how the management hierarchy affects it.
I find it uncomfortable to respond to what may be seen as a threat here, lets get back to the technical aspects.

**Comment by RamLavi** [comment]
As posting this test as an XFail has brought up maintainer opposition, currently closing this PR in order to incorporate this test in the future PR(s) that will fix the issue.

**Pull request closed without merging**
